### PR TITLE
Add apis for machine-class

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**: Adding APIs for Machine-class. Machineclass is a way of externalizing the ProviderConfig from MachineSpec using reference. 
This PR is basically re-initiating the work done here: https://github.com/kubernetes/kube-deploy/pull/659
- I will migrate the review-comments from that PR to here.

**Which issue(s) this PR fixes** : This PR partially resolves- #22 .

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Added apis for machine-class
```

### CODE REVIEW

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 1]
```diff
@@ -39,5 +42,25 @@ type ProviderConfig struct {
 // ProviderConfigSource represents a source for the provider-specific
 // resource configuration.
 type ProviderConfigSource struct {
```
Comment (hardikdr) [inline pkg/apis/cluster/v1alpha1/common_types.go]: mvladev: https://github.com/kubernetes/kube-deploy/pull/659/files#r177866614

roberthbailey: I don't think it should be necessary to specify the capacity / allocatable on a single machine if you inline the provider config. I think of the class is a separate way to bring the provider specific data, not something inherent to the spec of the machine.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 2]
```diff
@@ -39,5 +42,25 @@ type ProviderConfig struct {
...
-	// TODO(roberthbailey): Fill these in later
+	// The machine class from which the provider config should be sourced.
+	// +optional
+	MachineClass *MachineClassRef `json:machineClass,omitempty`
+}
+
+// MachineClassRef is a reference to the MachineClass object. Controllers should find the right MachineClass using this reference.
+type MachineClassRef struct {
```
Comment (hardikdr) [inline pkg/apis/cluster/v1alpha1/common_types.go]: mvladev: What about namespace?

roberthbailey: Storage classes aren't namespaced, and I used that same pattern here. But it's worth discussing whether they should be part of a namespace.

Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/common_types.go]: I have entertained the namespaces- should be helpful especially keeping in mind the sensitivity of the data in machine-classes and our approach of isolating the clusters based on namespaces.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 3]
```diff
@@ -39,5 +42,25 @@ type ProviderConfig struct {
...
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Parameters allow basic substitution to be applied to
+	// a MachineClass (where supported).
+	// Keys must not be empty. The maximum number of
+	// parameters is 512, with a cumulative max size of 256K.
+	// TODO(roberthbailey): Should this be a json-patch?
+	// +optional
+	Parameters map[string]string `json:parameters,omitempty`
```
Comment (hardikdr) [inline pkg/apis/cluster/v1alpha1/common_types.go]: krousey: What kinds of things is this meant to override? Capacity? Allocatable? Or stuff in the provider config? All of these seem like they could be deeply nested structures that a map would not be able to fully express.

If ProviderConfig, perhaps we consider json patches?

roberthbailey: My thinking was in the provider config (e.g. parameterize the zone of a class to stamp out identical machine sets in multiple zones).

File: pkg/apis/cluster/v1alpha1/machineclass_types.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	// How much capacity is actually allocatable on this machine.
+	// Must be equal to or less than the capacity, and when less
+	// indicates the resources reserved for system overhead.
+	//
+	// WARNING: It is up to the creator of the MachineClass to ensure that
+	// this field is consistent with the underlying machine that will
+	// be provisioned when this class is used, to inform higher level
+	// automation (e.g. the cluster autoscaler).
+	Allocatable corev1.ResourceList `json:"allocatable"`
```
Comment (hardikdr) [inline pkg/apis/cluster/v1alpha1/machineclass_types.go]: **krousey**: From what I understand, this is the capacity minus the amount that kubelet is going to reserve on the node. I don't think you can know what that reserve is going to be just given a machine class. I know for GKE, we vary the reserve for each release, and possibly by machine size.

Perhaps the best way to represent this would be to officially model the Kubelet reserved resources in the MachineSpec (and therefore MachineSet and MachineDeployment). If we did that, we could drop Allocatable here, and autoscalers of deployments could pick the capacity from machine class and subtract reserved from machine spec.

**mvladev**: I agree with @krousey. There is no way to calculate in advance the allocated resources - the Machine class doesn't have a knowledge of which kubelet version / container runtime you are going to use and those can affect the kubelet's --kube-reserved and --system-reserved flags.

**roberthbailey**: The intent is that this would indicate to the cluster autoscaler how much actual capacity would exist once the "node allocatable" overhead was subtracted. The fact that the overhead varies by version makes putting this variable here... difficult, since it would tightly couple a machineclass to a specific k8s version if you wanted to adhere to the warning text.

@krousey - Is your suggestion of putting reserved in the machine spec to put it next to the kubelet version, since it is tightly coupled with it?

File: pkg/apis/cluster/v1alpha1/machineclass_types.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+	// Provider-specific configuration to use during node creation.
+	ProviderConfig runtime.RawExtension `json:"providerConfig"`
+
+	// TODO: should this use an api.ObjectReference to a 'MachineTemplate' instead?
+	// A link to the MachineTemplate that will be used to create provider
+	// specific configuration for Machines of this class.
+	// MachineTemplate corev1.ObjectReference `json:machineTemplate`
```
Comment (hardikdr) [inline pkg/apis/cluster/v1alpha1/machineclass_types.go]: **mladev**: Why is this needed?

**roberthbailey**: I added this comment as an alternate design -- one thing we'd thought about was splitting this data across two objects (class + template) and the template could potentially be used directly by machine in addition to machine class. Not sure if that's valuable or not though, so I put this here to foster discussion.

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 1 above)
Comment (roberthbailey) [reply pkg/apis/cluster/v1alpha1/common_types.go]: /cc @mvladev - additional thoughts on this?

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 3 above)
Comment (roberthbailey) [reply pkg/apis/cluster/v1alpha1/common_types.go]: I think we should leave this out for now. If we find a need for it, we can figure out how to add it later (templating vs. overlays will be contentious). But until there is a concrete need, less is more.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 6]
```diff
@@ -39,5 +42,25 @@ type ProviderConfig struct {
...
-	// TODO(roberthbailey): Fill these in later
+	// The machine class from which the provider config should be sourced.
+	// +optional
+	MachineClass *MachineClassRef `json:machineClass,omitempty`
+}
+
+// MachineClassRef is a reference to the MachineClass object. Controllers should find the right MachineClass using this reference.
+type MachineClassRef struct {
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (roberthbailey) [inline pkg/apis/cluster/v1alpha1/common_types.go]: Isn't there a "reference" type that we should use instead of objectmeta? Something like LocalReference or ObjectReference?

File: pkg/apis/cluster/v1alpha1/machineclass_types.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineClass can be used to templatize and re-use provider configuration
+// across multiple Machines / MachineSets / MachineDeployments.
+// +k8s:openapi-gen=true
+// +resource:path=machineclasses
+type MachineClass struct {
```
Comment (roberthbailey) [inline pkg/apis/cluster/v1alpha1/machineclass_types.go]: If we leave out both allocatable and capacity, then machine class == provider config. I think it's important that we address @mvladev's comment above about whether we should have both of those fields on all machines (via an inline class) or only available to the autoscaler on machines that were created via a reference to a machine class.

File: sample/machineclass.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Sample machine-class object
+apiVersion: "cluster.k8s.io/v1alpha1"
+kind: MachineClass
+metadata:
+  name: small
+  namespace: foo
+capacity:
```
Comment (roberthbailey) [inline sample/machineclass.yaml]: leave out capacity / allocatable since those are currently commented out.

File: sample/machineclass.yaml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  storage: 20Gi
+allocatable:
+  cpu: 2
+  memory: 7.5Gi
+  storage: 20Gi
+providerConfig:
+  apiVersion: "gceproviderconfig/v1alpha1"
+  kind: "GCEProviderConfig"
+  project: "$GCLOUD_PROJECT"
+  zone: "${ZONE:-us-central1-f}"
```
Comment (roberthbailey) [inline sample/machineclass.yaml]: since we are going to remove the parameters part, make this just `us-central1-f` instead of my bash-style substitution syntax.

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 7 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: Agreed. I believe these decisions would be easier to take once requirements from cluster-autosclaer requirements are clearer. Shall we evolve the APIs for such fields as and when needed?
See: Initial work has started here https://github.com/kubernetes/community/pull/2653

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 6 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/common_types.go]: Yes, my bad, made it objectreference.

File: sample/machineclass.yaml (hunk 8 above)
Comment (hardikdr) [reply sample/machineclass.yaml]: Done. Thanks for pointing out, I missed it.

File: sample/machineclass.yaml (hunk 9 above)
Comment (hardikdr) [reply sample/machineclass.yaml]: Fixed.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 10]
```diff
@@ -39,5 +42,17 @@ type ProviderConfig struct {
...
+}
+
+// MachineClassRef is a reference to the MachineClass object. Controllers should find the right MachineClass using this reference.
+type MachineClassRef struct {
+	// +optional
+	*corev1.ObjectReference `json:",inline"`
+
+	// Provider is the name of the cloud-provider which MachineClass is intended for.
+	// +optional
+	Provider string `json:"provider,omitempty"`
```
[the remark this answers is not in the document]
Comment (roberthbailey) [reply pkg/apis/cluster/v1alpha1/common_types.go]: @mvladev - where does the machine class contain that info? Right now it's contained in this string.

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 7 above)
Comment (roberthbailey) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: We are going to chat with some folks that work on the autoscaler tomorrow; maybe that will help clarify what is needed here.

Comment (enxebre) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: Code on how allocatable is calculated is here https://github.com/kubernetes/kubernetes/blob/426ef9d349bb3a277c3e8826fb772b6bdb008382/pkg/kubelet/cm/node_container_manager.go#L174:33
We could assume a reasonable margin so allocatable is set based on the given capacity when creating a new machineClass. This would be used for scaling form zero, otherwise info from running machines/nodes would be used
Also should taints belongs to the machineClass?

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 2 above)
Comment (sflxn) [reply pkg/apis/cluster/v1alpha1/common_types.go]: StorageClass are basically templates.  Here, we're using namespace for isolating actual clusters.  Do we need to also isolate machine templates?  I don't see a need for namespace here.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 11]
```diff
@@ -39,5 +42,17 @@ type ProviderConfig struct {
 // ProviderConfigSource represents a source for the provider-specific
 // resource configuration.
 type ProviderConfigSource struct {
-	// TODO(roberthbailey): Fill these in later
+	// The machine class from which the provider config should be sourced.
+	// +optional
+	MachineClass *MachineClassRef `json:machineClass,omitempty`
```
Comment (sflxn) [inline pkg/apis/cluster/v1alpha1/common_types.go]: I just want to point out I believe this is different from the way StorageClass and RuntimeClass handles ref.  I guess it's ok since we're not handling the class the same way as them.  They use just a string.  Their controller matches the class' string name to match up the provider.  So if we were following the same pattern, we would have some controller that matched up the string with a provider known to the controller.  It's been a few months since I looked at this, but I believe that is the pattern.

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sflxn) [reply pkg/apis/cluster/v1alpha1/common_types.go]: I assume Provider is similar to StorageClass' "provisioner"?  If so, shouldn't we also have a name field for the MachineClass?  It only matters if we follow the same pattern as StorageClass and RuntimeClass.  I know there is now a DeviceClass that is being proposed too.  I think they're all trying to pattern after StorageClass.

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 7 above)
Comment (sflxn) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: It seems strange to have allocatable and capacity part of machineclass.  Is this defining how many pods we're going to allow to run on this machine?  How can we know that without knowing the size of OS, and additional software stack?  Shouldn't we let the control plane calculate that?

File: sample/machineclass.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+metadata:
+  name: test-machine 
+  namespace: foo
+  labels:
+    test-label: test-label 
+spec:
+  providerConfig:
+    valueFrom:
+      machineClass: 
+        provider: gcp
```
Comment (sflxn) [inline sample/machineclass.yaml]: I know this is an example, but can provider be defined in the machine class object above?  With the way it was defined above, can provider be provided in either machine class or machine objects?

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 2 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/common_types.go]: As mentioned, machine-classes going forward would be expected to have specific details towards machines and in-turn clusters, and also sensitive information of underlying infra. Separating machine-classes via namespaces will help to avoid unintended consumers of it.

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 7 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: Yes, these fields are expected to be available only after certain calculation, but autoscaler probably would be asking for it from a machine-api stack. We are in talks with autoscaler folks and try to decide right place/design, and hence I have intentionally kept those fields out for the first cut.

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/common_types.go]: Yes, ObjectReference should provide us with the name field in it.

File: sample/machineclass.yaml (hunk 12 above)
Comment (hardikdr) [reply sample/machineclass.yaml]: I have kept `provider` at machine-object level, keeping in mind the usecase where we might not want to fetch the complete MachineClass when MachineObject is seen/updated. In the future iterations, we can think if it's needed at any other layers as well to be populated explicitly.

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 7 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: @enxebre I am sure about the specific requirements around taints, but I would expect Taints to be available rather at MachineObjects than at class. Class should be seen more for representation of set of machines under MachineDeployment/Set, where subset of them could or could not have taints, which could be best identified by taints on specific MachineObjects.

**Comment by vikaschoudhary16** [comment]
Trying to compare machine classes with storage classes. Motivation behind storage classes was the portability across different storage providers. Wondering what is the motivation behind machine class? Can you please link me to a user story that how autoscaler will consume machine classes?

File: sample/machineclass.yaml (hunk 12 above)
Comment (hardikdr) [reply sample/machineclass.yaml]: Inline providerConfig will still work as it is currently, classes will only be a parallel mechanism for users to inject the providerConfig via reference. Above comment was explicitly for `provider` field at MachineObject layer, and not the existence of classes as such :)

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 1 above)
Comment (vikaschoudhary16) [reply pkg/apis/cluster/v1alpha1/common_types.go]: @roberthbailey does that mean machine class is just a collection of provider specific configuration? 
Advantages of machine classes:
* less configuration lines in the machine objects by keeping common bits in a machine class object

Any other?

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 2 above)
Comment (vikaschoudhary16) [reply pkg/apis/cluster/v1alpha1/common_types.go]: tend to agree with @hardikdr on keeping machine class as namespaced object.

File: pkg/apis/cluster/v1alpha1/common_types.go (hunk 1 above)
Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/common_types.go]: > less configuration lines in the machine objects by keeping common bits in a machine class object
- more precisely, externalize the `providerConfig` to separate objects.
- as discussed in wg-call, it could also be used for defaulting/versioning via apiserver-webhooks, which could be hard to achieve in only inlined.

**Comment by hardikdr** [comment]
@vikaschoudhary16 Though we learn from storage-class, we may have our divergent-points, as underlying usecase is different, for instance - namespaced machineclasses.
You might want to be updated with developments in this WIP-PR, I have described few points there
: https://github.com/kubernetes/community/pull/2653

File: pkg/apis/cluster/v1alpha1/machineclass_types.go (hunk 5 above)
Comment (vikaschoudhary16) [reply pkg/apis/cluster/v1alpha1/machineclass_types.go]: +1 for keeping `capacity` and `allocatable` as `MahcineStatus` fields. cluster-api controller would initialize these fields once linking with the kubelet node is established. These are "observed" values and mary vary from machine to machine due to some discovery failure or hw failure etc. Therefore should be updated in MachineStatus whatever is actually observed at kubelet node.

+1 for keeping taints in MachineClassSpec. Taints are generally used for scheduling domain isolation and therefore generally applies to a bunch of machines.

**Comment by roberthbailey** [comment]
@hardikdr - this needs a rebase (for the generated file).

**Comment by hardikdr** [comment]
@roberthbailey I updated the commits to align with both comment from @vikaschoudhary16 and kubebuilder pr, can you please have a look at the changes made.

**Comment by roberthbailey** [comment]
/lgtm
/approve

This has been waiting for comments for longe enough. Let's get it in and then we can tweak it going forward if we need to (since the api is still alpha).
